pkg/cluster/cluster.go

```diff
@@ -197,16 +187,37 @@ func (c *Clusterer) Start() error {
 	case *localNode:
 		return nil // no-op, always ready
 	case *GossipNode:
-		err := node.Start() // TODO(@tpaschalis) Should we backoff and retry before moving on to the fallback here?
+		peers, err := node.GetPeers()
+		if err != nil {
+			return err
+		}
+		err = node.Start(peers) // TODO(@tpaschalis) Should we backoff and retry before moving on to the fallback here?
 		if err != nil {
 			level.Debug(node.log).Log("msg", "failed to connect to peers; bootstrapping a new cluster")
-			node.cfg.JoinPeers = nil
-			err = node.Start()
+			err = node.Start(nil)
 			if err != nil {
 				return err
 			}
 		}
 
+		if node.cfg.RejoinInterval > 0 {
+			go func() {
+				t := time.NewTicker(node.cfg.RejoinInterval)
+				for {
+					<-t.C
+					peers, err := node.GetPeers()
+					if err != nil {
+						level.Error(node.log).Log("msg", "failed to refresh the list of peers", "err", err)
+						continue // we'll try on the next tick
+					}
+					err = node.Start(peers)
+					if err != nil {
+						level.Error(node.log).Log("msg", "failed to rejoin the list of peers", "err", err)
+					}
+				}
+			}()
+		}
```

Comment:
Tiny nit: this technically leaks and doesn't have a way to ever terminate the goroutine.

Reply:
Initially I'd passed the parent context to stop the ticker when it's cancelled, but when this happens we'd already be exiting. I'll fix it either way.